javascript/extractor/lib/typescript/src/type_table.ts

```diff
@@ -23,6 +23,28 @@ function isTypeVariable(type: ts.Type): type is ts.TypeVariable {
   return (type.flags & ts.TypeFlags.TypeVariable) !== 0;
 }
 
+/**
+ * Returns `true` if the properties of the given type can safely be extracted
+ * without restricting expansion depth.
+ *
+ * This predicate is very approximate, and considers all unions, intersections,
```

Comment:
Very approximate indeed. 
Maybe we should look into special casing on some common cases later.

Reply:
It shouldn't be that bad. The structure of a type is still extracted, this just determines whether its properties are. So for a type like `Buffer | string`, we can still tell it's a `Buffer | string`, but we won't necessarily compute the members of the union itself (i.e. the intersection of members on `Buffer` and `string`).

Also, members are always computed for types directly referenced from the AST, for example, if some expression is inferred to have the type `A & B`, that type is fully extracted, members and all. On the other hand, if `{ field: C & D }` is inferred, then that type is fully extracted, but the types `C` and `D` might be shallow, because it won't blindly unfold the `&` anymore.